### DISCUSSION TITLE
Test mode so we can test payment flows without actually paying with BTC or tBTC

### DISCUSSION THREAD

**Discussion by woutersamaey** [btcpayserver/btcpayserver#917] [Ideas / Feature Requests]
I am testing BTCPay Server integration with Magento 2 and was wondering if I can put BTCPay Server in "test mode" where I don't actually have to pay with BTC on the payment page.

Some payment providers allow you to put your store in "test mode" and a few buttons will appear on the payment page:
- Simulate full payment
- Simulate partial payment (not common for traditional credit card payments, but might be useful for crypto)
- Simulate payment failed

If you click on "Simulate full payment", the payment is considered OK and you proceed to the next step.
The other buttons can be similar and trigger other scenarios.

This is only for integration testing.

Another idea that might be handy is that instead of putting the store in test mode, the payment request itself indicates that it is a test request so in automated testing or end-to-end testing situations, one does not need to keep enabling and disabling test mode all the time.

Or, there could be a "test user" who is always in test mode while the rest is in production mode.

**Comment by pavlenex** [comment]
Hey @woutersamaey are you familiar with testnet and regtest?
Basically, you can run environment variable `export NBITCOIN_NETWORK="testnet"`
It will launch your server on testnet network which will look like [this](https://testnet.demo.btcpayserver.org/).

1. Set testnet network.
2. Get a testnet wallet and get some testnet coins via faucets online.
3. Pay and test
Small info on Testnet can be found [here](https://docs.btcpayserver.org/btcpay-basics/tryitout).

Not completely sure that's what you're looking for.

**Comment by woutersamaey** [comment]
@pavlenex I am familiar, but:
- It requires running a second BTCPay Server instance on another server (can't use Docker for running 2 instances on the same machine)
- Testers will need tBTC. In a larger organisation, testers may not be educated, will run out of tBTC, will need to share passwords, lose access to test wallets, etc. Additional overhead.
- Automated testing (Selenium) is not possible.

**Comment by Kukks** [comment]
How about just using Lightning to pay? creating 1sat invoices on mainnet will not break the bank and are instant anyway :)

**Comment by woutersamaey** [comment]
@Kukks for this we need to have test products set up to cost just 1 sat or 0,01 USD. Also takes time to change the prices, again the wallet problem, etc.

**Comment by pavlenex** [comment]
When you put it in that perspective, I see why you need it, I usually use LN as Kukks said, or just mark invoices as completed manually. Will mark this as a feature request, if you wish to expand more on it feel free to do so.

**Comment by woutersamaey** [comment]
@pavlenex I am willing to work on this, but I'm not a C# dev. I have installed VS Code on my Mac and can read and understand the code (looks a bit like Java Spring MVC to me).
The main issue I will face is best practises and logic issues. If someone is willing to coach me a bit, I'm happy to do it.

**Comment by pavlenex** [comment]
@woutersamaey Awesome, we do have a #developers chat on Mattermost at chat.btcpayserver.org where I'm sure some of the devs will be more than happy to help you out when you have the time join us and let's keep improving this, also thanks for your contributions so far.

**Comment by woutersamaey** [comment]
If I'm correct, the following scenario's are possible. Can someone confirm or add?
I want to compile a list of all possible scenarios a merchant can encounter, so we can create test flows and have BTCPay Server simulate those at the push of a button. For each of these, there should be a button on the "test payment" page.

I'd suggest making a "test payment" page that looks like the real payment page, except that the QR 
 and address should be replaced with these buttons. A different color or alert message would also help explaining this is not a production page.

An extra flag when the invoice is created on BTCPay Server can indicate the customer is going for a test payment.

**Before expiration**
- Simulate full payment (100%)
- Simulate minor underpayment (99%)
- Simulate minor overpayment (101%)
- Simulate big underpayment (80%)
- Simulate big overpayment (120%)

**After expiration**
- Simulate full payment after expiration (100%)
- Simulate minor underpayment after expiration (99%)
- Simulate minor overpayment after expiration (101%)
- Simulate big underpayment after expiration (80%)
- Simulate big overpayment after expiration (120%)

**Non-payment**
- Simulate abandonment (user leaves the page)
- Simulate payment failed (what does this look like?)


Anything else you can think of that a merchant would want to have test flows for?
Some of these may seam a bit extreme, but serious merchants WILL ask for these in my experience.

**Comment by pavlenex** [comment]
@woutersamaey all of those situations look okay. Some indeed look a bit extreme. 

For
Non-payment

    * Simulate abandonment (user leaves the page) - invoice simply expires without being paid
    * Simulate payment failed (what does this look like?) - I think the term here for the invoice `invalid` but the status it triggers can be `failed`  

**Invalid**
Paid, but failed to receive a sufficient amount of confirmations within the time specified in-store settings
Check the transaction on a blockchain explorer, if it received sufficient confirmations, mark as complete

**Comment by royserpa** [comment]
> 
> 
> When you put it in that perspective, I see why you need it, I usually use LN as Kukks said, or just mark invoices as completed manually. Will mark this as a feature request, if you wish to expand more on it feel free to do so.

Is it possible to manually mark invoices as all possible variables? It's to debug and so on. If it can be done, how can I do it? Thanks!

**Comment by pavlenex** [comment]
@royserpa You can manually change **certain** states by clicking on a single invoice drop-down in Invoices > Status column.
![Screen Shot 2019-08-16 at 11 58 38](https://user-images.githubusercontent.com/36959754/63160126-4ffbd700-c01d-11e9-8c72-f288e9c22697.png)

**Comment by royserpa** [comment]
> 
> 
> @royserpa You can manually change **certain** states by clicking on a single invoice drop-down in Invoices > Status column.
> ![Screen Shot 2019-08-16 at 11 58 38](https://user-images.githubusercontent.com/36959754/63160126-4ffbd700-c01d-11e9-8c72-f288e9c22697.png)

You are right, I saw that. But I don't think it's possible to mark invoices as paid, confirmed, complete :/

**Comment by royserpa** [comment]
> 
> 
> > @royserpa You can manually change **certain** states by clicking on a single invoice drop-down in Invoices > Status column.
> > ![Screen Shot 2019-08-16 at 11 58 38](https://user-images.githubusercontent.com/36959754/63160126-4ffbd700-c01d-11e9-8c72-f288e9c22697.png)
> 
> You are right, I saw that. But I don't think it's possible to mark invoices as paid, confirmed, complete :/

Without sending any btc/tbtc or lightning btc :/

**Comment by pavlenex** [comment]
It is not possible without sending anything. That's I think what this issue is about. 

But payments flow is way better tested in regtest and dev environment where you don't have to spend anything.

**Comment by danielcharrua** [comment]
@woutersamaey I have the same issue as you. I use a WooCommerce store but it will be great to have some switch in BTCPay plugin to test payments. Any news on this? Have you managed to solve it?
Thanks.

**Comment by woutersamaey** [comment]
@danielcharrua no this issue is still open.

In development, you can run a command on the command line to fake a payment.
A solution might be to run a 2nd install in regtest (developer mode) and run this CLI command to fake a payment automatically. It’s not an easy setup, but if you want fully automated end-to-end tests, this would be a way forward.

@pavlenex an easier to use test payment would still be helpful, but I guess it would require huge refactoring?

**Comment by Kukks** [comment]
With greenfield, you can use an api command to mark an invoice as complete.
Would that be enough?

**Reply by woutersamaey** [reply]
Sure! But ideally for testing, there should be a tester-friendly way to trigger each of the invoice states (and the 6 different webhooks), so the merchant can test his business flow for every occasion. 

Traditional payment processors do this using a special test checkout page with some buttons or ask you to use a specific email or fake visa card for each case. 

Ie if you use visa card 4111 1111 1111 1111 it's deemed a successful payment. Another card number will trigger declined etc. 

In the case of BTCPay Server, there are more cases because of over or underpayment. Making this harder to test and harder on the merchant.

**Reply by danielcharrua** [reply]
I agree with @woutersamaey on e-commerce stores, we usually have a sandbox URL to test payments, e.g. when testing Stripe on WooCommerce we have a testing mode to create some fake payments and see if they arrive ok to Stripe, also test different scenarios like "declined card", "accepted card", etc..

This way we can test in a "developer mode" that all transactions are arriving to BTCPay and how is being handled. After all tests we can turn into production as we already know that our testing payments are passing ok.

I know BTCPay Server has already a self-hosted instance running on Testnet but this is not running on our instance, maybe those testing payments are running ok on your instance but not on ours.

**Comment by danielcharrua** [comment]
Here you can see how Stripe for example in WooCommerce handles the test mode with different API keys for each `test` & `prod` environments. The same behavior as Stripe is all across other payment plugins, e.g. PayPal, etc...

![Captura de pantalla 2021-06-27 a las 17 42 11](https://user-images.githubusercontent.com/31829061/123550847-b13f7600-d76f-11eb-8d72-6e403ef0ef80.jpg)

**Comment by woutersamaey** [comment]
@NicolasDorier @Kukks would it be technically feasible to:
- Add an "is_test" boolean to the invoice model
- Change the checkout page for test invoices to contain a few "simulate payment" buttons
- Create a "fake" payment in the DB so the invoice appears as under, exact or overpaid? The fake payments would be flagged as fake obviously.

I imagine this is quite the refactor, but would it work if the DB is altered this way?
Willing to look into this if you believe it's doable.

Mostly asking because I have no idea if this is a DB manipulation or if payments are loaded / synced from the blockchain, making this more difficult.

**Reply by danielcharrua** [reply]
@Kukks any updates on this? Thank you.

**Reply by Kukks** [reply]
I'm not sure I like this approach as it adds quite a bit of complexity for a niche feature. I would suggest that we go towards implementing a more general-purpose solution. My suggestion would be to revisit https://github.com/btcpayserver/btcpayserver/pull/920 as this would allow your use case. Maybe we can reduce scope by making it API only to reduce controversy and get consensus

**Reply by woutersamaey** [reply]
I agree @Kukks that this is a niche feature, but we should have some way to test more easily. On regtest we have CLI commands that simulate a payment. Can we build on this to meet this goal? Maybe do a double deployment with mainnet and regtest in parallel? Could these 2 share resources perhaps?

Just brainstorming here. If it doesn't help, just say so.

**Reply by danielcharrua** [reply]
@Kukks I see that a more general-purpose solution can be a good approach. 

Other solutions as Bitpay (I'm not comparing this to BTCPay 🙌 only for illustrating), have also a WooCommerce plugin and in the plugin configuration you can use a _Development API key_ or a _Production API key_, to test both scenarios. [Please see the second image on this link](https://bitpay.com/integrations/woocommerce)

Maybe we can use the API only method you are talking about to let e-commerce developers and makers to do some tests interacting with BTCPay Server to see that is all working as it should. Simulating some payments and scenarios.
I think this is the goal here, what do you think @woutersamaey? Thank you both.

**Reply by woutersamaey** [reply]
@danielcharrua I just proposed a PR #2672 that moves us in the right direction.

**Reply by danielcharrua** [reply]
Great work @woutersamaey I'm looking onto the PR.

**Reply by woutersamaey** [reply]
Please note that this is only part of the solution. You'd need to run a server on regtest.
Next steps would be:
- A docker deployment that contains both a mainnet + regtest server
- Tooling to copy production data to development server so both can be identical (or close to identical with some data stripped or anonymized). You could then periodically and automatically sync your mainnet server (production) to your regtest server (testing).

**Reply by danielcharrua** [reply]
This seems a lot to do and not an easy setup. If we know that the BTCPay Server works ok managing BTC payments, why are we also testing this?

Maybe an easier path would be only testing the e-commerce platform communication with BTCPay Server and fake some payments between BTCPay and the e-commerce (Magento, WooCommerce, etc). I that way we do not need to test anything related with maninnet, regtest or testnet. Let me know what do you think.

**Reply by woutersamaey** [reply]
We're on the same page. I just need regtest, so we can test payments without actually moving coins. Regtest has a bottomless wallet. By going this route, we avoid overhauling the software and leverage our existing development/testing tools for shopkeepers to use.

We also want to simulate the movement of coins, so we can trigger the many scenarios like multiple, partial, over and underpayment. Something that is never an issue with other payment providers as it is with bitcoin and other cryptos. This creates more and special circumstances a business needs to deal with.

The deployment could be made simple for the merchant if we provide an optional fragment for that.
The syncing of data would be optional and only for advanced users.

The reason why we're not faking communication is that both users of the shop + accounting people can verify the results in the system they know and have conform that everything is as they would expect.

**Reply by danielcharrua** [reply]
@woutersamaey any updates/news on this? have you managed to create and run a test environment?

**Reply by woutersamaey** [reply]
Yes. This is available. It's called "cheat mode" and only works on regtest + you need to enable it.

Installation is a bit advanced so I can't get into it right now.

**Reply by danielcharrua** [reply]
are there any plans for writing a guide or similar on the future? For other users to use this feature.

**Reply by pavlenex** [reply]
@danielcharrua you basically only need to deploy btcpayserver locally https://docs.btcpayserver.org/Development/LocalDevelopment/ after that, create a store, a wallet, and then in Wallet > Receive tab, you'll see "cheat mode". Click on that to generate a bunch of transactions to your wallet. Then you can go create an invoice. On the invoice itself in regtest mode, you'll see a button to mine the blocks or you can manually pay that invoice.

**Reply by woutersamaey** [reply]
This works on a dev system.

If you want to have this on a server with Docker, configuration is different. If you're looking to setup a server that runs continuously, let me know and I'll look it up on my server. I have a setup like this running.

**Comment by n-hadi** [comment]
A sandbox function is desperately needed still.

**Reply by woutersamaey** [reply]
This has been implemented already. You just need to run an instance on regtest instead of mainnet or testnet. 
In regtest, there are buttons on the checkout page to fake a payment and to fake mine 1 or more blocks.

**Reply by n-hadi** [reply]
alright, thanks. ill look into it

**Reply by n-hadi** [reply]
yeah I have looked into it and its a pain in the ass. I have let LunaNode automatically deploy a BTCPay Server instance for me since Im not a DevOps engineer and have no idea about all this Docker shit.
It would  be nice if you could simply mark a regular invoice as "test invoice" through the api.

**Reply by JakeCirino** [reply]
> This has been implemented already. You just need to run an instance on regtest instead of mainnet or testnet. In regtest, there are buttons on the checkout page to fake a payment and to fake mine 1 or more blocks.

Saying "this has been implemented already" is just a lazy cop-out when your "implementation" is forcing the user to edit env vars or run a private instance of the server just to mock test payments.

**Comment by alexandervlpl** [comment]
Looks like the regtest setup is separate from the the standard Docker installation, it's clearly intended for developing BTCPay Server itself. Looking at the docs it's not clear if and how I can put my existing BTCPay Server in "regtest" mode with this additional network running locally. Maybe a guide is needed? Does anyone have some tips?

From my perspective I just need to test my simple Python integration against a standard server. The free testnet coins are so hard to get, people are selling them for real money. :rofl: I'm still waiting for my 5 cents. This is really way more complicated than it should be, we're missing functionality that's standard in any payment service as mentioned above. Pretty huge gap in what is otherwise a great open source solution!

Waiting for real payments to clear can also be very slow for development/testing.

**Reply by woutersamaey** [reply]
Run the docker setup, but instead of using "mainnet", just put "regtest". The rest of the docker setup is the same.

You should also add a Docker variable "cheat_mode" set to "true"

That should be all you need.

**Reply by alexandervlpl** [reply]
I must be doing something wrong, still don't have cheat mode after this:

Generated/docker-compose.generated.yml:
```
btcpayserver:
    environment:
        CHEAT_MODE: "true"
```
Updated setup:
```
export NBITCOIN_NETWORK="regnet"
. ./btcpay-setup.sh -i
```
Also, Jesus... I think it's syncing an entire node which will take days and terabytes...
```
2024-05-31T06:49:07Z UpdateTip: new best=0000000000000000005f68b3f1aefd0a23fa23feecf60fce03c96318d46b323c height=305610 version=0x00000002 log2_work=79.144157 tx=40642915 date='2014-06-13T17:07:56Z' progress=0.039930 cache=831.1MiB(7414796txo)
```

**Reply by woutersamaey** [reply]
It's not `regnet`, but `regtest`
From the docs
```
NBITCOIN_NETWORK: The type of network to use (eg. mainnet, testnet, or regtest. Default: mainnet)
```
And cheatmode should be (iirc)
```
cheatmode=1 (or true, I'm not sure. You probably should not be using "true" but true without quotes as that would make it a string instead of a boolean).
```
 Sorry, no underscore. Cheat mode allows you do to a fake payment from the pay page. At the bottom of the pay page a form appears to make a fake payment + fake mine a few blocks. Cheatmode only works in combination with `regtest` of course. If you have regtest, but no cheatmode, you can still fake payments, but you'll need to manually create a transaction from another store wallet or run commands from CLI. Example: mining a few blocks would be a CLI command. Cheatmode is way more convenient though.

@NicolasDorier Maybe we should add `cheatmode` to the Docker docs? I see it's not mentioned and very useful for those looking to develop integrations.

@alexandervlpl can you let me know if you got cheatmode working? I'll look into adding it to the docs.

**Reply by alexandervlpl** [reply]
Thanks @woutersamaey , that typo was my first problem.

I found the cheat setting in the code, it's `BTCPAY_CHEATMODE: "true"` with quotes like the other Docker environment variables. This is definitely applied, because now the server crashes immediately in `Cheater.cs`. Do I need more developer settings to get this working?

```
Unhandled exception. System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at NBitcoin.RPC.RPCClient.SendCommandAsync(RPCRequest request, CancellationToken cancellationToken)
   at NBitcoin.RPC.RPCClient.SetVersion(RPCCapabilities capabilities)
   at NBitcoin.RPC.RPCClient.ScanRPCCapabilitiesAsync(CancellationToken cancellationToken)
   at BTCPayServer.Services.Cheater.Microsoft.Extensions.Hosting.IHostedService.StartAsync(CancellationToken cancellationToken) in /source/BTCPayServer/Services/Cheater.cs:line 63
   at Microsoft.AspNetCore.Hosting.HostedServiceExecutor.ExecuteAsync(Func`2 callback, Boolean throwOnFirstFailure)
   at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.StartWithTasksAsync(IWebHost webHost, CancellationToken cancellationToken) in /source/BTCPayServer/Extensions/WebHostExtensions.cs:line 23
   at BTCPayServer.Program.Main(String[] args) in /source/BTCPayServer/Program.cs:line 76
   at BTCPayServer.Program.Main(String[] args) in /source/BTCPayServer/Program.cs:line 103
   at BTCPayServer.Program.<Main>(String[] args)
```

**Reply by woutersamaey** [reply]
Hmmm... only thing I can think of it that when you change any of the ENV variables, you need to re-run the btcpay-setup bash script, I believe.
It's been a while, so I could be wrong.
There shouldn't be anything else needed. You're definately very close.

Maybe @NicolasDorier can help you out further or go ask someone on Mattermost chat support channel.

**Reply by alexandervlpl** [reply]
I nuked it from space, put `BTCPAY_CHEATMODE` in a Docker fragment as [documented](https://docs.btcpayserver.org/Docker/#how-can-i-customize-the-generated-docker-compose-file) and did a completely clean install. Same thing, `CHEATMODE` causes this authorization problem. :disappointed: Will keep looking.

**Comment by robwoodgate** [comment]
What would be most user-friendly would be:

1. An option in Store settings to put the store in "Test Mode"
2. Enabling above would then surface more admin invoice options, allowing invoice to be simulated into the [various test states](https://github.com/btcpayserver/btcpayserver/discussions/917#discussioncomment-202203) - eg: ability to set additional status conditions like partially paid, overpaid etc
3. On disabling store "Test Mode" option, invoices used for testing could be archived / expired or otherwise as needed to "clean up"

This would allow testing without actually making live payments... in the same way live invoices can be marked as invalid or settled manually.

This way, users can create a store specifically for testing integrations etc and delete it afterwards if they wish.

**Reply by woutersamaey** [reply]
Faking payments (as it is currently) is also tied to fake mining blocks and the events happening on the blockchain. This is why this is only available on regtest, where these things are fakeable as well.
